### PR TITLE
Don't raise an error if stdout/stderr are closed

### LINKED ISSUE #1116: BrokenPipeError leading to a crash when using filtering/counting shell utilities
When using e.g. tail or head to limit the number of lines a BrokenPipeError is thrown and leads to an abnormal exit of borg:

```
% sudo borg list BACKUP | head -n1
20160202-121810                      Tue, 2016-02-02 13:03:51
Local Exception.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 1498, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 1452, in run
    return args.func(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 80, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 544, in do_list
    print(format_archive(archive_info))
BrokenPipeError: [Errno 32] Broken pipe

Platform: Linux occam 4.6.0-bait.0-wbt #1 SMP PREEMPT Tue May 17 09:32:06 CEST 2016 x86_64 
Linux: debian stretch/sid 
Borg: 1.0.2  Python: CPython 3.5.1+
PID: 21174  CWD: /home/user
sys.argv: ['/usr/bin/borg', 'list', 'BACKUP']
SSH_ORIGINAL_COMMAND: None

Exception ignored in: <_io.TextIOWrapper name='<stdout>' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
```

***

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/34723987-brokenpipeerror-leading-to-a-crash-when-using-filtering-counting-shell-utilities)

### PR DESCRIPTION
**Pull request by PlasmaPower**
Fixes #1116

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
stdout might transport file data (borg extract --stdout), so I don't think we want to silently suppress errors there in general.

**Comment by PlasmaPower** [comment]
@ThomasWaldmann That's the only case though, correct? `borg extract` is also a read-only command, so if it fails it's not a big deal. What if someone wants to do `borg extract --stdout repo::example file.log | tail -n 100`? If it's `borg extract --stdout repo::example file.log > file.log`, and it runs out of disk space, then Bash will output an error saying that it's out of disk space. Whatever's on the other end of the pipe should output the error causing the closing of the pipe if it was closed unintentionally.

**Comment by PlasmaPower** [comment]
Maybe `ErrorFree` is the wrong term, because it can still through a `ValueError` if the stream is already closed. I haven't seen that cause any problems though.

File: src/borg/helpers.py [hunk 1]
```diff
@@ -1546,3 +1547,27 @@ def signal_handler(signo, handler):
...
+        return ''
+
+    def write(self, s):
+        if not self.closed:
+            try:
+                return super(ErrorFreeTextIOWrapper, self).write(s)
+            except IOError:
+                try:
+                    super(ErrorFreeTextIOWrapper, self).close()
+                except IOError:
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: One minor nitpick, quoting from the python 3.4 docs:
```
The following exceptions are kept for compatibility with previous versions; starting from Python 3.3, they are aliases of OSError.

exception EnvironmentError

exception IOError

exception WindowsError
```

So, please use OSError at all these places.

File: src/borg/helpers.py [hunk 2]
```diff
@@ -1546,3 +1547,27 @@ def signal_handler(signo, handler):
         yield
     finally:
         signal.signal(signo, old_signal_handler)
+
+
+class ErrorFreeTextIOWrapper(io.TextIOWrapper):
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: Maybe ErrorIgnoring... ?

File: src/borg/helpers.py [hunk 3]
```diff
@@ -1546,3 +1547,27 @@ def signal_handler(signo, handler):
...
+    def read(self, n):
+        if not self.closed:
+            try:
+                return super(ErrorFreeTextIOWrapper, self).read(n)
+            except IOError:
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: shall we maybe ignore a bit more specifically?
like only the closed pipe error?

**Comment by PlasmaPower** [comment]
So I'm now catching OSError, should I only catch `errno.EPIPE`?

**Comment by PlasmaPower** [comment]
Okay, I'm now only catching `errno.EPIPE` but my code has sort of turned into a mess of conditionals. Not much can be done about it though.

**Comment by PlasmaPower** [comment]
Checking if it's not equal instead simplifies thing a great deal. I think this is ready to merge, unless anyone has any additional comments.

File: src/borg/helpers.py [hunk 4]
```diff
@@ -1546,3 +1548,33 @@ def signal_handler(signo, handler):
...
+        if not self.closed:
+            try:
+                return super(ErrorIgnoringTextIOWrapper, self).read(n)
+            except OSError as err:
+                if err.errno != errno.EPIPE:
+                    raise
+                try:
+                    super(ErrorIgnoringTextIOWrapper, self).close()
+                except OSError as err:
+                    if err.errno != errno.EPIPE:
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: can this happen?

File: src/borg/helpers.py [hunk 5]
```diff
@@ -1546,3 +1548,33 @@ def signal_handler(signo, handler):
...
+        if not self.closed:
+            try:
+                return super(ErrorIgnoringTextIOWrapper, self).write(s)
+            except OSError as err:
+                if err.errno != errno.EPIPE:
+                    raise
+                try:
+                    super(ErrorIgnoringTextIOWrapper, self).close()
+                except OSError as err:
+                    if err.errno != errno.EPIPE:
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: same here

File: src/borg/helpers.py (hunk 4 above)
Comment (PlasmaPower) [inline src/borg/helpers.py]: I'm not sure. It's probably be safer just to ignore any OSError when closing anyway, the only reason I'm closing is to clean up. I've changed this.

**Comment by ThomasWaldmann** [comment]
i don't think we have unit tests for this right?
so, did you test the latest code manually?

**Comment by PlasmaPower** [comment]
@ThomasWaldmann Yep I tested `borg info repo::archive` and `borg info repo::archive | head -n 1`.

**Comment by ThomasWaldmann** [comment]
ok, code lgtm - merge?

just one thing: #1116 is in 1.0.7 milestone and as we usually merge 1.0-maint into master anyway, PR against 1.0-maint would be easier to process.

**Comment by PlasmaPower** [comment]
@ThomasWaldmann Sounds good - I'll close then then reapply the patch to 1.0-maint.

**Pull request closed without merging**